docs/design/coreclr/botr/data_contracts_for_coreclr.md

```diff
@@ -0,0 +1,430 @@
+
+### Global Values
+Global values which can be of types (int8, uint8, int16, uint16, int32, uint32, int64, uint64, pointer, unicode string)
+All global values have a unicode string describing their name, and a value of one of the above types.
+
+### Compatible Contract
+Each compatible contract is described by a unicode string naming the contract, and a uint32 version. It is an ERROR if multiple versions of a contract are specified in the stream.
+
+### Data Structure Layout
+Each data structure layout has a unicode name for the type, followed by a list of fields. These fields can be of primitive types (int8, uint8, int16, uint16, int32, uint32, int64, uint64, pointer) or of another named data structure type. Each field descriptor provides the offset of the field, the name of the field, and the type of the field.
+
+## Versioning of contracts
+Contracts are described an integer version number. A higher version number is not more recent, it just means different. In order to avoid conflicts, all contracts should be documented in the main branch of the dotnet repository with a version number which does not conflict with any other. It is expected that every version of every contract describes the same functionality/data layout/set of global values.
+
+## Contract data model
+Logically a contract may refer to another contract. If it does so, it will typically refer to other contracts by names which do not include the contract version. This is to allow for version flexibility. Logically once the contract data stream is fully processed, there is a single list of contracts that represents the set of contracts useable with whatever runtime instance is being processed.
+
+## Types of contracts
+
+There are 3 different types of contracts each representing a different phase of execution of the data contract system.
+
+### Composition contracts
+These contracts indicate the version numbers of other contracts. This is done to reduce the size of contract list needed in the runtime data stream. In general it is intended that as a runtime nears shipping, the product team can gather up all of the current versions of the contracts into a single magic value, which can be used to initialize most of the contract versions of the data contract system. A specific version number in the contract data stream for a given contract will override any composition contracts specified in the contract data stream. If there are multiple composition contracts in a stream which specify the same contract to have a different version, the first composition contract linearly in the runtime data stream wins. This is intended to allow for a composite contract for the architecture/os indepedent work, and a separate composite contract for the non independent work. If a contract is specified explicitly in the Data Contract Stream and a different version is specified via the composition contract mechanism, the explicitly specified contract takes precedence.
```

Comment:
💯 I think it's important to have composition contracts as a component of the descriptor.

Reply:
My preference right now is that there would be relatively few versioned contracts, each of which represents all the data structures and algorithms necessary to do a swath of related diagnostic functionality. For example contracts might be:
- Enumerate all threads and detail information about each
- Enumerate all loaded assemblies and provide details about each
- Describe the state of the GC heap and enumerate all objects on it

Right now I don't see the value in finer grained contracts but I do see costs - such as needing the complexity of managing composition contracts.

> Are they intended to be logical groupings?

I don't know if David envisioned the same thing, but yes I think of them as logical groupings.

Reply:
Logical groupings are what I see as sensible. My sense of taste wants to have more contracts than fewer, but there is a cost to that which folks keep pointing out. My preference would be to do fine grained contracts but really only use the composition contracts to keep things sane, but I see the general agreement is that that is really complex to manage over time, so I think we can also just do fewer big contracts.